daphne/src/dap_error.rs

```diff
@@ -0,0 +1,68 @@
+use std::fmt::Display;
+
+use crate::{aborts::DapAbort, messages::TransitionFailure, vdaf::VdafError};
+
+/// DAP errors.
+#[derive(Debug, thiserror::Error)]
+pub enum DapError {
+    /// Fatal error. If this triggers an abort, then treat this as an internal error.
+    ///
+    /// To create an instance of this variant the [`fatal_error`] must be used. This ensures that
+    /// all fatal errors are logged with tracing when created.
+    #[error("fatal error: {0}")]
+    Fatal(#[from] FatalDapError),
+
+    /// Error triggered by peer, resulting in an abort.
+    #[error("abort: {0}")]
+    Abort(#[from] DapAbort),
+
+    /// Transition failure. This error blocks processing of a paritcular report and may, under
+    /// certain conditions, trigger an abort.
+    #[error("transition error: {0}")]
+    Transition(#[from] TransitionFailure),
+}
+
+impl FatalDapError {
+    #[doc(hidden)]
+    pub fn __use_the_macro(s: String) -> Self {
+        FatalDapError(s)
+    }
+}
+
+impl From<VdafError> for DapError {
+    fn from(e: VdafError) -> Self {
+        match e {
+            VdafError::Codec(..) | VdafError::Vdaf(..) => {
+                Self::Transition(TransitionFailure::VdafPrepError)
+            }
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct FatalDapError(String);
+
+impl std::error::Error for FatalDapError {}
+
+impl Display for FatalDapError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+#[macro_export]
```

Comment:
Why is a macro needed? I wonder if we couldn't just add a method on `DapError` (or maybe replace `DapError::fatal()`) with something that does the same thing as this code.

The reason I would prefer not to use a macro (or at least not this one) Is that, in use, the syntax looks relatively complex. For example:

```rust
fatal_error!(
                err = format!("{e:?}"), // `HpkeError` doesn't implement Display or Error :(
                ?kem_id,
                "bad key generation for KEM",
            )
```
The `err = ` and `?kem_id` look really strange to someone who is not as expert at Rust as you (me :D)

Reply:
Maybe I should have specified the design goals of the macro, I'm not 100% happy with it either but I pushed the PR because I wanted feedback.

I wanted the first argument to it to be an implementation of `std::error::Error`, sadly if I require that `&str` and `String` aren't allowed and that just breaks a lot of current code, so I added the `err =` thing to force the caller to think about it a bit and pass in the error in situations like `.map_err(|e| ...)` where you do have an error.

The `?kem_id` part is just how tracing works and so it felt like something okay to introduce, since this is effectively a thin wrapper around `tracing` anyway.

Finally the big reason I want this is:
- it forces the `fatal_error` field on the log line, which is easy to filter for in kibana
- it makes the line numbers happen exactly where the error is generating, making it really easy to trace back to the origin

Reply:
> ... and that just breaks a lot of current code ...

Does this indicate to you that there is a desirable code design pattern that we are not following (consistently)? If so, it might be worth taking the time to fix/refactor the code such that it does.

I don't object to this macro if you think it's the right thing to do. The main consideration for me is that we do right thing rather than the fast thing.